syntax = "proto3";

package spq_generated;

service SortingPriorityQueueService {
  rpc Dequeue(DequeueRequest) returns (ItemResponse) {}
  rpc Peek(PeekRequest) returns (ItemResponse) {}
  rpc GetSize(GetSizeRequest) returns (GetSizeResponse) {}
  rpc Enqueue(EnqueueRequest) returns (EnqueueResponse) {}
  rpc GetEpoch(GetEpochRequest) returns (GetEpochResponse) {}
  rpc CreateQueue(CreateQueueRequest) returns (QueueResponse) {}
}

message Feature {
  int64 value = 1;
  string name = 2;
}

message EnqueueRequest {
  bytes item = 1;
  repeated Feature features = 2;
  string queueName = 3;
}

message EnqueueResponse {
  int64 size = 1;
}

message GetSizeRequest {
  string queueName = 1;
}

message GetSizeResponse {
  int64 size = 1;
}

message PeekRequest {
  string queueName = 1;
}

message DequeueRequest {
  string queueName = 1;
}

message ItemResponse {
  bytes item = 1;
  bool hasItem = 2;
  int64 size = 3;
}

message GetEpochRequest {
  string queueName = 1;
}

message GetEpochResponse {
  int64 epoch = 1;
}

enum Type {
  IN_MEMORY = 0;
  DURABLE = 1;
}

message CreateQueueRequest {
  string name = 1;
  Type queueType = 2;
  repeated string features = 3;
}

message QueueResponse {
  string name = 1;
}

service HealthService {
  rpc Check(HealthCheckRequest) returns (HealthCheckResponse);

  rpc Watch(HealthCheckRequest) returns (stream HealthCheckResponse);
}

message HealthCheckRequest {
  string service = 1;
}

message HealthCheckResponse {
  enum ServingStatus {
    UNKNOWN = 0;
    SERVING = 1;
    NOT_SERVING = 2;
  }
  ServingStatus status = 1;
}
